util/file_reader_writer.cc

```diff
@@ -514,6 +514,31 @@ class ReadaheadRandomAccessFile : public RandomAccessFile {
     return Status::OK();
   }
 
+  virtual Status ReadaheadBackwards(uint64_t offset) {
+    Status s;
+    size_t prefetch_offset = Roundup(offset, alignment_);
+    if (prefetch_offset < prefetch_bytes) {
+      prefetch_offset = 0;
+    } else {
+      prefetch_offset -= prefetch_bytes;
+    }
+    if (prefetch_offset == buffer_offset_) {
+      return s;
+    }
+    Slice readahead_result;
+    s = file_->Read(prefetch_offset, prefetch_bytes, &readahead_result,
+                    buffer_.BufferStart());
```

Comment:
We should not have two places to read data from file_. Try to define a helper function and call it in both of here and line 487.